src/main/scala/diplomacy/LazyModule.scala

```diff
@@ -112,25 +167,42 @@ abstract class LazyModule()(implicit val p: Parameters)
     children.filter(!_.omitGraphML).foreach { c => c.edgesGraphML(buf, pad) }
   }
 
+  /** Execute function by all of this node's [[children]].
+    *
+    * @param iterfunc function to run
+    * */
   def nodeIterator(iterfunc: (LazyModule) => Unit): Unit = {
     iterfunc(this)
     children.foreach( _.nodeIterator(iterfunc) )
   }
 
+  /** Accessor of val [[children]]. */
   def getChildren = children
 }
 
 object LazyModule
 {
+  /** Current [[LazyModule]] scope. The scope is a stack of [[LazyModule]]/[[LazyScope]].
+    * Each call to [[LazyScope.apply]] or [[LazyModule.apply]] will push that item onto the current scope.
+    * */
   protected[diplomacy] var scope: Option[LazyModule] = None
+  /** Global index of [[LazyModule]], notice there is no zeroth module. */
   private var index = 0
 
+  /** Any instance of [[LazyModule]] should be wrapper in to a [[LazyModule.apply]],
+    * Making use this singleton manage scope and index for elaboration.
+    *
+    * @param bc [[LazyModule]] instance that needs to be wrapped.
+    * @param valName [[ValName]] used to name this instance,
+    *                it can be automatically generated by [[ValName]] macro, or specify manually.
+    * @param sourceInfo [[SourceInfo]] information about where this [[LazyModule]] is being generated
+    * */
   def apply[T <: LazyModule](bc: T)(implicit valName: ValName, sourceInfo: SourceInfo): T = {
-    // Make sure the user put LazyModule around modules in the correct order
-    // If this require fails, probably some grandchild was missing a LazyModule
-    // ... or you applied LazyModule twice
-    require (scope.isDefined, s"LazyModule() applied to ${bc.name} twice ${sourceLine(sourceInfo)}")
+    /** Make sure the user puts [[LazyModule]] around modules in the correct order. */
```

Comment:
```suggestion
    // Make sure the user puts [[LazyModule]] around modules in the correct order.
```

I'm changing any of the /** ... */ comments that really should be // comments back to //